### PR TITLE
Batch write imported stuff in importmulti

### LINKED ISSUE #15739: large ranged descriptor imports are sloooooooooow.
Performing the following (note the range of 10,000):

`importmulti '[{"desc": "sh(wpkh([aeee1e6a/49h/1h/0h]tpubDC5bVdaveQYxGFztMShFAjU637kKjrs9FD8ne1oHR7z8tLeQJExsaxWRVeSpcCcYMdp5ecf2LQhuVaeYFZ7YEpvWTJr4nDGUJrmuvLuTSsp/1/*))", "keypool": true, "timestamp": "now", "watchonly": true, "internal": true, "range": {"end": 1000}}]' '{"rescan": false}'`

takes many minutes. @achow101 reports a similar import taking 6 minutes on his machine.

### PR DESCRIPTION
**Pull request by achow101**
Instead of writing each item to the wallet database individually, do them in batches so that the import runs faster.

This was tested by importing a ranged descriptor for 10,000 keys.

Current master

```
$ time src/bitcoin-cli -regtest -rpcwallet=importbig importmulti '[{"desc": "sh(wpkh([73111820/44h/1h/0h]tpubDDoT2SgEjaU5rerQpfcRDWPAcwyZ5g7xxHgVAfPwidgPDKVjm89d6jJ8AQotp35Np3m6VaysfUY1C2g68wFqUmraGbzhSsMF9YBuTGxpBaW/1/*))#3w7php47", "range": [0, 10000], "timestamp": "now", "internal": true, "keypool": false, "watchonly": true}]'
...

real	7m45.29s
```

This PR:

```
$ time src/bitcoin-cli -regtest -rpcwallet=importbig4 importmulti '[{"desc": "pkh([73111820/44h/1h/0h]tpubDDoT2SgEjaU5rerQpfcRDWPAcwyZ5g7xxHgVAfPwidgPDKVjm89d6jJ8AQotp35Np3m6VaysfUY1C2g68wFqUmraGbzhSsMF9YBuTGxpBaW/1/*)#v65yjgmc", "range": [0, 10000], "timestamp": "now", "internal": true, "keypool": false, "watchonly": true}]'
...

real	3.93s
```

Fixes #15739

### CODE REVIEW

**Comment by gwillen** [comment]
utACK although it would be good to factor out the counting into a WalletWriteCache or something, that flushes itself at intervals so you don't have to do it manually.

File: src/wallet/rpcdump.cpp [hunk 1]
```diff
@@ -1282,33 +1287,46 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
             }
              const CPubKey& pubkey = entry->second;
              CPubKey temp;
-             if (!pwallet->GetPubKey(id, temp) && !pwallet->AddWatchOnly(GetScriptForRawPubKey(pubkey), timestamp)) {
+             if (!pwallet->GetPubKey(id, temp) && !pwallet->AddWatchOnlyWithDB(*batch, GetScriptForRawPubKey(pubkey), timestamp)) {
```
Comment (gwillen) [inline src/wallet/rpcdump.cpp]: There is a weird indentation glitch here, although it looks like it was preexisting.

File: src/wallet/rpcdump.cpp [hunk 2]
```diff
@@ -1282,33 +1287,46 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
...
+                    batch.reset(new WalletBatch(pwallet->GetDBHandle()));
+                }
             }
             CTxDestination dest;
             ExtractDestination(script, dest);
             if (!internal && IsValidDestination(dest)) {
                 pwallet->SetAddressBook(dest, label, "receive");
             }
         }
+        batch.reset();
```
Comment (gwillen) [inline src/wallet/rpcdump.cpp]: Is this necessary? It's about to go out of scope, right?

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (Empact) [reply src/wallet/rpcdump.cpp]: What are you referring to? I don't see it.

Comment (Empact) [reply src/wallet/rpcdump.cpp]: Ah, think I see it now - 1283-1287 are off by one.

**Comment by Empact** [comment]
https://github.com/Empact/bitcoin/commit/dd5b0aaf58d843c2532b3ca1829bf90482e9dc52 `CWallet::AddKeyOrigin(const CPubKey& pubkey, const KeyOriginInfo& info)` and `CWallet::WriteKeyMetadata(const CKeyMetadata& meta, const CPubKey& pubkey, const bool overwrite)` can be removed.

https://github.com/Empact/bitcoin/commit/8896384a6c04eb75dcd01742adfb84bde76e9e46 How about moving this functionality into CWallet, so that we don't operate externally on CWallet::database any more than we already do. As the `GetDBHandle` comments say: "Ideally this function would not be necessary.". Practically speaking, this also explicitly limits the scope of the batch operations on `database` to within the lifecycle of the `CWallet` instance.

File: src/wallet/wallet.h [hunk 3]
```diff
@@ -1222,6 +1224,7 @@ class CWallet final : public CCryptoKeyStore, private interfaces::Chain::Notific
 
     /** Add a KeyOriginInfo to the wallet */
     bool AddKeyOrigin(const CPubKey& pubkey, const KeyOriginInfo& info);
+    bool AddKeyOriginWithDB(WalletBatch& batch, const CPubKey& pubkey, const KeyOriginInfo& info);
```
Comment (Empact) [inline src/wallet/wallet.h]: nit: May be a bad example to follow, but `AddKeypoolPubkeyWithDB` has `batch` as the last argument, which makes as much sense to me.

**Comment by Empact** [comment]
Concept ACK

**Review by promag — COMMENTED** [review]
Concept ACK, nice improvement.

**Comment by MarcoFalke** [comment]
See also:  

* Use a single wallet batch for UpgradeKeyMetadata #15433

**Comment by achow101** [comment]
I've pulled in @Empact's changes (Empact@dd5b0aa was squashed into 59ec41257f37f2a0438b10ef9776bf7ea3488e9a, cherry-picked the other two commits).

File: src/wallet/rpcdump.cpp [hunk 4]
```diff
@@ -1263,17 +1263,22 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
...
              const CKeyID& id = entry.first;
              assert(key.VerifyPubKey(pubkey));
              pwallet->mapKeyMetadata[id].nCreateTime = timestamp;
              // If the private key is not present in the wallet, insert it.
-             if (!pwallet->HaveKey(id) && !pwallet->AddKeyPubKey(key, pubkey)) {
+             if (!pwallet->HaveKey(id) && !pwallet->AddKeyPubKeyWithDB(*batch, key, pubkey)) {
                  throw JSONRPCError(RPC_WALLET_ERROR, "Error adding key to wallet");
              }
              pwallet->UpdateTimeFirstKey(timestamp);
+             if (++cnt % 1000 == 0) {
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/rpcdump.cpp]: The batch can get very large in memory if lots of keys are involved, so we limit the size to avoid blowing up memory during imports.

File: src/wallet/wallet.h (hunk 3 above)
Comment (achow101) [reply src/wallet/wallet.h]: I was following the `AddKeyPubKeyWithDB` way of argument ordering.

File: src/wallet/rpcdump.cpp (hunk 2 above)
Comment (achow101) [reply src/wallet/rpcdump.cpp]: Probably not. I did that just as a belt-and-suspenders thing.

**Comment by gwillen** [comment]
Especially if there are now multiple places we're manually batching writes like this, and more where we're considering it: Let me suggest again and more strongly that the batch object should be handling autoflushing after every X calls, instead of open-coding it in the caller like this. If it's appropriate for every caller, it could go into WalletBatch, else into a new class that's an extremely thin wrapper around it. I bet the total number of lines added in this PR would actually go down.

**Comment by Sjors** [comment]
Concept ACK. I tried the first two commits (so without address book change) with #14145 and importing 2x 1000 keys seems about 6 times faster.

**Comment by Empact** [comment]
@gwillen Agree it should be done but would be appropriate as a follow-up

**Comment by practicalswift** [comment]
Looks like `AddKeypoolPubkey` is no longer used and can be removed?

**Comment by achow101** [comment]
I've implemented @gwillen's suggestion. `WalletBatch` will flush to disk automatically after every 1000 writes. This appears to be faster too as it does not close and reopen the database on every flush. With this change, I've also updated `UpgradeKeyMetadata`.

> Looks like `AddKeypoolPubkey` is no longer used and can be removed?

Done

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -389,15 +388,10 @@ void CWallet::UpgradeKeyMetadata()
...
-                batch->WriteKeyMetadata(meta, pubkey, true);
-                if (++cnt % 1000 == 0) {
-                    // avoid creating overlarge in-memory batches in case the wallet contains large amounts of keys
-                    batch.reset(new WalletBatch(*database));
-                }
+                batch.WriteKeyMetadata(meta, pubkey, true);
             }
         }
     }
-    batch.reset(); //write before setting the flag
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: in commit ebb0ea85539b3e2f515fca89231c20ac86fb52d5:

Why is this no longer needed?

Comment (achow101) [reply src/wallet/wallet.cpp]: When batch goes out of scope, it will write. This was just a belt and suspenders.

**Comment by achow101** [comment]
I'm unsure why tests are currently failing.

File: src/wallet/wallet.cpp (hunk 5 above)
Comment (gwillen) [reply src/wallet/wallet.cpp]: The comment suggests that in this case -- unlike the batch write in importmulti -- there's a specific reason to force it to write before the end of the scope.

File: src/wallet/walletdb.h [hunk 6]
```diff
@@ -157,6 +157,9 @@ class WalletBatch
             return false;
         }
         m_database.IncrementUpdateCounter();
+        if (m_database.nUpdateCounter % 1000 == 0) {
```
Comment (promag) [inline src/wallet/walletdb.h]: I get that this makes the code simple, but IMO this defeats the batch purpose. Like, it shouldn't be something to worry about here.

File: src/wallet/db.cpp [hunk 7]
```diff
@@ -614,7 +614,9 @@ void BerkeleyBatch::Flush()
     if (fReadOnly)
         nMinutes = 1;
 
-    env->dbenv->txn_checkpoint(nMinutes ? gArgs.GetArg("-dblogsize", DEFAULT_WALLET_DBLOGSIZE) * 1024 : 0, nMinutes, 0);
+    if (env) {
```
Comment (promag) [inline src/wallet/db.cpp]: Why this change?

File: src/wallet/walletdb.h (hunk 6 above)
Comment (gwillen) [reply src/wallet/walletdb.h]: I don't understand what you mean by 'defeats the batch purpose'. Are there cases where the caller is depending on the batch not being flushed?

I think it is far better engineering practice to deal with periodic flushing once, inside the batch-management code, than to deal with it independently at each callsite. If there's some situation where periodic flushing is undesirable, I think it would be better to offer both versions.

File: src/wallet/db.cpp (hunk 7 above)
Comment (achow101) [reply src/wallet/db.cpp]: So that tests pass. When there's a dummy database, env will be `nulltpr` and this causes a segfault in tests which use a dummy database.

File: src/wallet/wallet.cpp (hunk 5 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: reverted

File: src/wallet/db.cpp (hunk 7 above)
Comment (promag) [reply src/wallet/db.cpp]: Can't we make `env` is defined instead? Otherwise a comment would make sense as this doesn't really happen-

File: src/wallet/walletdb.h (hunk 6 above)
Comment (promag) [reply src/wallet/walletdb.h]: @gwillen I mean you should be able to create a batch with 1001 changes no? I was suggesting something on top of `WalletBatch` to do that.

Comment (gwillen) [reply src/wallet/walletdb.h]: Well, you can add as many changes as you want, it just flushes after every 1000. This should be completely transparent to the user. Unless there is some reason that flushing can hurt, like the caller is using the batch as though it were an atomic transaction?

Comment (achow101) [reply src/wallet/walletdb.h]: > Unless there is some reason that flushing can hurt, like the caller is using the batch as though it were an atomic transaction?

There is only one place where an atomic transaction is used, and if an atomic transaction is active, then `Flush()` will do nothing.

File: src/wallet/db.cpp (hunk 7 above)
Comment (achow101) [reply src/wallet/db.cpp]: > Can't we make `env` is defined instead?

No. `env` is nullptr for dummy databases which are used in tests. A check like this is used in many other places in this file in the form of `BerkeleyDatabase::IsDummy()`.

> Otherwise a comment would make sense as this doesn't really happen.

Done.

Comment (Empact) [reply src/wallet/db.cpp]: nit: could deal with it earlier, and use `IsDummy` to make it more explicit.

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -390,10 +389,6 @@ void CWallet::UpgradeKeyMetadata()
             CPubKey pubkey;
             if (GetPubKey(meta_pair.first, pubkey)) {
                 batch->WriteKeyMetadata(meta, pubkey, true);
-                if (++cnt % 1000 == 0) {
-                    // avoid creating overlarge in-memory batches in case the wallet contains large amounts of keys
-                    batch.reset(new WalletBatch(*database));
-                }
```
Comment (Empact) [inline src/wallet/wallet.cpp]: nit: could squash this into the previous

File: src/wallet/wallet.cpp [hunk 9]
```diff
@@ -3404,13 +3404,6 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
     return true;
 }
 
-void CWallet::AddKeypoolPubkey(const CPubKey& pubkey, const bool internal)
-{
-    WalletBatch batch(*database);
-    AddKeypoolPubkeyWithDB(pubkey, internal, batch);
-    NotifyCanGetAddressesChanged();
-}
-
```
Comment (Empact) [inline src/wallet/wallet.cpp]: nit: could squash into https://github.com/bitcoin/bitcoin/pull/15741/commits/c088227bd612920f0e2ea3a34d6e5037490d62c2

File: src/wallet/walletdb.h (hunk 6 above)
Comment (Empact) [reply src/wallet/walletdb.h]: How about add an optional argument to `WalletBatch` that indicates the # of updates to flush after? This makes the operation opt-in/explicit and also allows for optionally adjusting the flushes to the number of operations - e.g. 2x600 rather than 1000 and 200, if that's desirable.

Comment (promag) [reply src/wallet/walletdb.h]: Well I may be wrong but I see the batch as an atomic transaction. Making 1500 updates and then aborting the batch should not commit 1000 updates IMO.

Comment (achow101) [reply src/wallet/walletdb.h]: > Well I may be wrong but I see the batch as an atomic transaction. Making 1500 updates and then aborting the batch should not commit 1000 updates IMO.

Except that's not how the batch works normally. Batches are not atomic unless you explicitly make them so by starting a transaction with the underlying `BerkeleyBatch`.  In that case, then the flushing won't do anything. In the normal case, each update in a batch itself is atomic and should be expected to be written to the wallet file. There is no aborting. As soon as a batch goes out of scope, every single update is committed to the file. If you make 1500 updates, the first 1000 will be written after you do the 1000th write operation, and then the last 500 will be written when the batch object is deleted.

File: src/wallet/wallet.cpp (hunk 8 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Done

File: src/wallet/wallet.cpp (hunk 9 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Done

File: src/wallet/db.cpp (hunk 7 above)
Comment (achow101) [reply src/wallet/db.cpp]: `IsDummy` is done by `BerkeleyDatabase`, not `BerkeleyBatch` so is not available to do here (although the check is completely identical).

**Review by Sjors — CHANGES_REQUESTED** [review]
tACK 396d065 modulo notification bug (inline)

I'm particularly happy that this moves some complexity out of the RPC codebase.

Importing 1000 pubkeys on an iMac with 3,4 GHz Intel Core i5 goes from 25 to 6 seconds.

**Comment by promag** [comment]
@achow101 looks like UnsetWalletFlag should reuse the batch instance:
Screenshot 2019-05-18 at 10 45 19
Please test/banchmark with 76c3906d751442ddb1b9bf1662d3e2effb53e950 (https://github.com/promag/bitcoin/tree/pr15741), at least here it gives a significant improvement.

**Comment by jonasschnelli** [comment]
Concept ACK

**Comment by jb55** [comment]
@promag's patch brings my import speeds down from 8 minutes to about 10 seconds

File: src/wallet/walletdb.h (hunk 6 above)
Comment (achow101) [reply src/wallet/walletdb.h]: I believe that is out of date

Comment (achow101) [reply src/wallet/walletdb.h]: I've updated the comment.

File: src/wallet/rpcdump.cpp [hunk 10]
```diff
@@ -1288,20 +1289,20 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
...
-             if (!pwallet->GetPubKey(id, temp) && !pwallet->AddWatchOnly(GetScriptForRawPubKey(pubkey), timestamp)) {
+             if (!pwallet->GetPubKey(id, temp) && !pwallet->AddWatchOnlyWithDB(batch, GetScriptForRawPubKey(pubkey), timestamp)) {
                 throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
             }
             pwallet->mapKeyMetadata[id].nCreateTime = timestamp;
 
             // Add to keypool only works with pubkeys
             if (add_keypool) {
-                pwallet->AddKeypoolPubkey(pubkey, internal);
+                pwallet->AddKeypoolPubkeyWithDB(pubkey, internal, batch);
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/rpcdump.cpp]: Done

**Comment by achow101** [comment]
I've pulled in @promag's patch.

**Comment by promag** [comment]
@achow101 please update times in OP.

**Comment by achow101** [comment]
I noticed that importing scripts would also be slow so I've added `AddCScriptWIthDB` in order to apply the same batching performance boost to such imports.

Also updated OP.

**Comment by jb55** [comment]
tACK 87f7befca04a095a06af003f9fe20d2e3c155d2f sooo fast. tested a rescan with my all my trezor keys from HWI and it worked great. thanks @achow101, hw wallet + core is actually usable now!

**Comment by promag** [comment]
Restarted travis, it was failing with 😕
```
47/121 - rpc_psbt.py failed, Duration: 5 s
stdout:
2019-05-18T17:46:43.670000Z TestFramework (INFO): Initializing test directory /tmp/test_runner_₿_🏃_20190518_173555/rpc_psbt_71
2019-05-18T17:46:47.538000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/test_framework.py", line 194, in main
    self.run_test()
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/rpc_psbt.py", line 156, in run_test
    assert_raises_rpc_error(-22, "", self.nodes[0].converttopsbt, signedtx['hex'])
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/util.py", line 105, in assert_raises_rpc_error
    assert try_rpc(code, message, fun, *args, **kwds), "No exception raised"
AssertionError: No exception raised
```

**Comment by promag** [comment]
utACK 87f7bef. nit, ampersand should be before space.

File: src/wallet/wallet.h [hunk 11]
```diff
@@ -1204,6 +1204,7 @@ class CWallet final : public CCryptoKeyStore, private interfaces::Chain::Notific
 
     /** Unsets a single wallet flag */
     void UnsetWalletFlag(uint64_t flag);
+    void UnsetWalletFlagWithDB(WalletBatch& batch, uint64_t flag);
```
Comment (Empact) [inline src/wallet/wallet.h]: nit: could be `private`

File: src/wallet/wallet.h [hunk 12]
```diff
@@ -872,6 +883,7 @@ class CWallet final : public CCryptoKeyStore, private interfaces::Chain::Notific
     //! Adds an encrypted key to the store, without saving it to disk (used by LoadWallet)
     bool LoadCryptedKey(const CPubKey &vchPubKey, const std::vector<unsigned char> &vchCryptedSecret);
     bool AddCScript(const CScript& redeemScript) override;
+    bool AddCScriptWithDB(WalletBatch& batch, const CScript& script);
```
Comment (Empact) [inline src/wallet/wallet.h]: nit: could be `private`

**Comment by Empact** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/15741/commits/87f7befca04a095a06af003f9fe20d2e3c155d2f

File: src/wallet/wallet.h (hunk 11 above)
Comment (achow101) [reply src/wallet/wallet.h]: Done

File: src/wallet/wallet.h (hunk 12 above)
Comment (achow101) [reply src/wallet/wallet.h]: Done

**Comment by jb55** [comment]
utACK 0db94e5

**Comment by ariard** [comment]
Tested ACK 0db94e5

Drop import time for 10000 keys ranged descriptor from 495.57s to 3.626s with 2.8 GHz Intel Core i7.

**Comment by Empact** [comment]
re-utACK https://github.com/bitcoin/bitcoin/pull/15741/commits/0db94e55dcbbfc741df463c404818d9033b4fff1 only change is re the privacy of `UnsetWalletFlagWithDB` and `AddCScriptWithDB`.
